src/sbvr-api/actions.ts

```diff
@@ -0,0 +1,141 @@
+
+export type ODataActionRequest = Omit<ParsedODataRequest, 'odataQuery'> & {
+	odataQuery: Omit<ODataQuery, 'property'> & {
+		property: {
+			resource: string;
+		};
+	};
+};
+
+type ActionReq = Express.Request;
+
+export type ODataActionArgs<Vocab extends string> = {
+	request: ODataActionRequest;
+	tx: Tx;
+	api: (typeof api)[Vocab];
+	id: Params['id'];
+	req: ActionReq;
+};
+export type ODataAction<Vocab extends string = string> = (
+	args: ODataActionArgs<Vocab>,
+) => Promise<Response>;
+
+const actions = {} as {
+	[vocab: string]: {
+		[resourceName: string]: {
+			[actionName: string]: ODataAction;
+		};
+	};
+};
+
+export const isActionRequest = (
+	request: ParsedODataRequest,
+): request is ODataActionRequest => {
+	// OData actions must always be POST
+	// See: https://www.odata.org/blog/actions-in-odata/
+	return (
+		request.method === 'POST' &&
+		request.odataQuery.property?.resource != null &&
+		actions?.[request.vocabulary]?.[request.resourceName]?.[
+			request.odataQuery.property.resource
+		] != null
+	);
+};
+
+export const runAction = async (
+	request: ODataActionRequest,
+	req: ActionReq,
+): Promise<Response> => {
+	if (api[request.vocabulary] == null) {
+		throw new BadRequestError();
+	}
+
+	const actionName = request.odataQuery.property.resource;
+	const action =
+		actions?.[request.vocabulary]?.[request.resourceName]?.[actionName];
+	if (!action) {
+		throw new BadRequestError();
+	}
+
+	return sbvrUtils.db.transaction(async (tx) => {
+		// in practice, the parser does not currently allow actions without a key
+		// e.g. /vocab/resource(...)/action compiles but
+		// /vocab/resource/action won't even compile, so (current) reality is that if canAcccess
+		// does not throw, there will be an id, but I kept the interface as optional id
+		// for when the parser has support for root level actions
+		const id = request.odataQuery.key
+			? await canRunAction(request, req, actionName, tx)
+			: undefined;
+
+		const applicationApi = api[request.vocabulary].clone({
+			passthrough: { tx, req },
+		});
+
+		return await action({
+			request: request,
+			tx,
+			api: applicationApi,
+			req: req,
+			id,
+		});
+	});
+};
+
+export const addAction = <Vocab extends string>(
+	vocabulary: Vocab,
+	resourceName: string,
+	actionName: string,
+	action: ODataAction<Vocab>,
+) => {
+	actions[vocabulary] ??= {};
+	actions[vocabulary][resourceName] ??= {};
+	actions[vocabulary][resourceName][actionName] = action;
+};
+
+export const canRunAction = async (
+	request: ParsedODataRequest,
+	req: ActionReq,
+	actionName: string,
+	tx: Tx,
+) => {
+	const canAccessUrl = request.url
+		.slice(1)
+		.replace(new RegExp(`(${actionName})$`), 'canAccess');
+
+	const applicationApi = api[request.vocabulary];
+	if (!applicationApi) {
+		throw new BadRequestError(`Could not find model ${request.vocabulary}`);
+	}
+
+	const res = await api[request.vocabulary].request({
+		method: 'POST',
+		url: canAccessUrl,
+		body: { action: actionName },
```

Comment:
This is one of the most interesting parts IMHO. We can actually ensure you only run the action if you have permission to it (see here for example of how it looks like). When we have filtered canAccess this can be further improve to run the filtered canAccess so that you can have full permissions over the action execution